fastestimator/dataset/batch_dataset.py

```diff
@@ -0,0 +1,132 @@
+                 datasets: Union[FEDataset, Iterable[FEDataset]],
+                 num_samples: Union[int, Iterable[int]],
+                 probability: Optional[Iterable[float]] = None):
+        self.datasets = to_list(datasets)
+        self.num_samples = to_list(num_samples)
+        self.probability = to_list(probability)
+        self.index_maps = [list(range(len(dataset))) for dataset in self.datasets]
+        self.same_feature = False
+        self._check_input()
+
+    def _check_input(self):
+        for num_sample in self.num_samples:
+            assert isinstance(num_sample, int) and num_sample > 0, "only accept positive integer type as num_sample"
+        # check dataset keys
+        dataset_keys = [set(dataset[0].keys()) for dataset in self.datasets]
+        for key in dataset_keys:
+            assert key, "found no key in datasets"
+        is_same_key = all([dataset_keys[0] == key for key in dataset_keys])
+        is_disjoint_key = sum([len(key) for key in dataset_keys]) == len(set.union(*dataset_keys))
+        if len(self.datasets) > 1:
+            assert is_same_key != is_disjoint_key, "dataset keys must be all same or all disjoint"
+        self.same_feature = is_same_key
+        if self.probability:
+            assert self.same_feature, "keys must be exactly same among datasets when using probability distribution"
+            assert len(self.datasets) == len(self.probability), "the length of dataset must match probability"
+            assert len(self.num_samples) == 1, "num_sample must be scalar for probability mode"
+            assert len(self.datasets) > 1, "number of datasets must be more than one to use probability mode"
+            assert sum(self.probability) == 1, "sum of probability must be 1"
+            for p in self.probability:
+                assert isinstance(p, float) and p > 0, "must provide positive float for probability distribution"
+        else:
+            assert len(self.datasets) == len(self.num_samples), "the number of dataset must match num_samples"
+        if not self.same_feature:
+            assert len(set(self.num_samples)) == 1, "the number of samples must be the same for disjoint features"
+
+    def _do_split(self, splits: Sequence[Iterable[int]]):
+        # Overwriting the split() method instead of _do_split
+        raise AssertionError("This method should not have been invoked. Please file a bug report")
+
+    def split(self, *fractions: Union[float, int, Iterable[int]]) -> Union['UnpairedDataset', List['UnpairedDataset']]:
+        new_datasets = [to_list(ds.split(*fractions)) for ds in self.datasets]
+        num_splits = len(new_datasets[0])
+        new_datasets = [[ds[i] for ds in new_datasets] for i in range(num_splits)]
+        results = [BatchDataset(ds, self.num_samples, self.probability) for ds in new_datasets]
+        # Re-compute personal variables
+        self.index_maps = [list(range(len(dataset))) for dataset in self.datasets]
```

Comment:
use max(len(datasets)) for each and later put the mod on the outside

Reply:
like this?